# The 'center' of the Torah

## Table of content <a class="anchor" id="TOC"></a>

* <a href="#bullet1">1 - Introduction</a>
* <a href="#bullet2">2 - Load Text-Fabric app and data</a>
* <a href="#bullet3">3 - Performing the queries</a>
* <a href="#bullet3x1">&nbsp;&nbsp;&nbsp;3.1 - Center word</a>
* <a href="#bullet3x2">&nbsp;&nbsp;&nbsp;3.2 - Center verse</a>
* <a href="#bullet3x3">&nbsp;&nbsp;&nbsp;3.3 - Center sentence</a>
* <a href="#bullet3x4">&nbsp;&nbsp;&nbsp;3.4 - Center clause</a>
* <a href="#bullet3x5">&nbsp;&nbsp;&nbsp;3.5 - Center phrase</a>
* <a href="#bullet3x6">&nbsp;&nbsp;&nbsp;3.6 - Center chapter</a>
* <a href="#bullet3x7">&nbsp;&nbsp;&nbsp;3.7 - Center word - alternative 1</a> 
* <a href="#bullet3x8">&nbsp;&nbsp;&nbsp;3.8 - Center word - alternative 2</a> 


# 1 - Introduction <a class="anchor" id="bullet1"></a>
##### [Back to TOC](#TOC)

What is the center of the Torah? By what measure should this be determined?

# 2 - Load Text-Fabric app and data <a class="anchor" id="bullet2"></a>
##### [Back to TOC](#TOC)

In [1]:
%load_ext autoreload
%autoreload 2

In [2]:
# Loading the Text-Fabric code
# Note: it is assumed Text-Fabric is installed in your environment.
from tf.fabric import Fabric
from tf.app import use

In [3]:
# load the BHSL app and data
BHS = use ("etcbc/BHSA",hoist=globals())

**Locating corpus resources ...**

Note: Thefeature documentation can be found at [ETCBC GitHub](https://github.com/ETCBC/bhsa/blob/master/docs/features/0_home.md) 

In [62]:
# The following will push the Text-Fabric stylesheet to this notebook (to facilitate proper display with notebook viewer)
BHS.dh(BHS.getCss())

# 3 - Performing the queries <a class="anchor" id="bullet3"></a>
##### [Back to TOC](#TOC)

The feature uses will be 'number'. This starts with 1. The manner of numbering objects differs per object type. The following are of interest for this research:

type | numbering
--- | ---
phrase_atom | within the book
clause_atom | within the book
sentence_atom | within the book
word | within the book

Note: Full feature documentation is found [here](https://github.com/ETCBC/bhsa/blob/master/docs/features/number.md)

**Warning: the impact of borderconditions still needs to be sorted out here. ALso still need to investigate why result 0 returns last word of Deuteronomy 34:12 (i.e. the last entry of the results). Hence, the results below are provisional.**

## 3.1 - Center word <a class="anchor" id="bullet3x1"></a>
##### [Back to TOC](#TOC)

In [78]:
# number of words in Torah (WARNING: as per ETCBC definition!) 
WordQuery = '''
book book=Genesis|Exodus|Leviticus|Numeri|Deuteronomium
  word 
'''

WordResults = BHS.search(WordQuery)

  0.46s 112927 results


In [54]:
F.otype.sInterval('word')

(1, 426590)

In [55]:
# start + delta: 1 + int(112927/2) = 1 + 56463 = 56464
T.sectionFromNode(56464)

('Leviticus', 8, 21)

In [56]:
T.text(56464)

'בַּ'

In [80]:
BHS.show(WordResults,start=56463,end=56463,multiFeatures=True)

## 3.2 - Center verse <a class="anchor" id="bullet3x2"></a>
##### [Back to TOC](#TOC)

In [63]:
# number of verses in Torah
VerseQuery = '''
book book=Genesis|Exodus|Leviticus|Numeri|Deuteronomium
  verse 
'''

VerseResults = BHS.search(VerseQuery)

  0.04s 5853 results


In [30]:
F.otype.sInterval('verse')

(1414389, 1437601)

In [34]:
# start + delta: 1414389 + int(5853/2) = 1414389 + 2926 = 1417315
T.sectionFromNode(1417315)

('Leviticus', 8, 9)

In [35]:
T.text(1417315)

'וַיָּ֥שֶׂם אֶת־הַמִּצְנֶ֖פֶת עַל־רֹאשֹׁ֑ו וַיָּ֨שֶׂם עַֽל־הַמִּצְנֶ֜פֶת אֶל־מ֣וּל פָּנָ֗יו אֵ֣ת צִ֤יץ הַזָּהָב֙ נֵ֣זֶר הַקֹּ֔דֶשׁ כַּאֲשֶׁ֛ר צִוָּ֥ה יְהוָ֖ה אֶת־מֹשֶֽׁה׃ '

In [65]:
BHS.show(VerseResults,start=2926,end=2926)

## 3.3 - Center sentence <a class="anchor" id="bullet3x3"></a>
##### [Back to TOC](#TOC)

In [40]:
# number of sentences in Torah
SentenceQuery = '''
book book=Genesis|Exodus|Leviticus|Numeri|Deuteronomium
  sentence 
'''

SentenceResults = BHS.search(SentenceQuery)

  0.13s 15088 results


In [42]:
F.otype.sInterval('sentence')

(1172308, 1236024)

In [43]:
# start + delta: 1172308 + int(15088/2) = 1172308 + 7544 = 1179852
T.sectionFromNode(1179852)

('Exodus', 36, 11)

In [44]:
T.text(1179852)

'וַיַּ֜עַשׂ לֻֽלְאֹ֣ת תְּכֵ֗לֶת עַ֣ל שְׂפַ֤ת הַיְרִיעָה֙ הָֽאֶחָ֔ת מִקָּצָ֖ה בַּמַּחְבָּ֑רֶת '

In [64]:
BHS.show(SentenceResults,start=7544,end=7544)

## 3.4 - Center clause <a class="anchor" id="bullet3x4"></a>
##### [Back to TOC](#TOC)

In [36]:
# number of clauses in Torah
ClauseQuery = '''
book book=Genesis|Exodus|Leviticus|Numeri|Deuteronomium
  clause 
'''

ClauseResults = BHS.search(ClauseQuery)

  0.11s 21181 results


In [37]:
F.otype.sInterval('clause')

(427559, 515689)

In [38]:
# start + delta: 427559 + int(21181/2) = 427559 + 10590 = 438149
T.sectionFromNode(438149)

('Leviticus', 4, 35)

In [39]:
T.text(438149)

'וְאֶת־כָּל־חֶלְבָּ֣ה יָסִ֗יר '

In [60]:
BHS.show(ClauseResults,start=10590,end=10590)

## 3.5 - Center phrase <a class="anchor" id="bullet3x5"></a>
##### [Back to TOC](#TOC)

In [48]:
# number of phrases in Torah
PhraseQuery = '''
book book=Genesis|Exodus|Leviticus|Numeri|Deuteronomium
  phrase 
'''

PhraseResults = BHS.search(PhraseQuery)

  0.47s 64195 results


In [49]:
F.otype.sInterval('phrase')

(651573, 904775)

In [50]:
# start + delta: 651573 + int(64195/2) = 651573 + 32097 = 683670
T.sectionFromNode(683670)

('Leviticus', 4, 32)

In [51]:
T.text(683670)

'נְקֵבָ֥ה תְמִימָ֖ה '

In [61]:
BHS.show(PhraseResults,start=32097,end=32097)

## 3.6 - Center chapter <a class="anchor" id="bullet3x6"></a>
##### [Back to TOC](#TOC)

In [45]:
# number of chapters in Torah
ChapterQuery = '''
book book=Genesis|Exodus|Leviticus|Numeri|Deuteronomium
  chapter 
'''

ChapterResults = BHS.search(ChapterQuery)

  0.03s 187 results


In [46]:
F.otype.sInterval('chapter')

(426630, 427558)

In [47]:
# start + delta: 426630 + int(187/2) = 426630 + 93 = 426723
T.sectionFromNode(426723)

('Leviticus', 4)

In [77]:
BHS.show(ChapterResults,start=93,end=93)

## 3.7 Center word - alternative 1<a class="anchor" id="bullet3x7"></a>
##### [Back to TOC](#TOC)

First check what can be placed after a word

In [127]:
# note: this is for the full TeNaCH!
Fs('trailer').freqList()

((' ', 236930),
 ('', 121801),
 ('&', 42275),
 ('00 ', 20146),
 ('05 ', 2266),
 ('00_S ', 1892),
 ('00_P ', 1165),
 ('_S ', 76),
 (' 05 ', 17),
 ('_P ', 13),
 ('00_N ', 7),
 ('00_N_P ', 1),
 ('00_N_S ', 1))

In [128]:
# Number of words in Torah (defined as items separeted by spaces OR maqaf - diacritical mark indicating a strong connection between words) 
# Following regexp (.+) selects for values of feature trailer that are 1 or more characters in length

WordQuery2 = '''
book book=Genesis|Exodus|Leviticus|Numeri|Deuteronomium
  word trailer~.+
'''

WordResults2 = BHS.search(WordQuery2)

  0.65s 79886 results


Find the midpoint: 79886/2 = 39948

In [121]:
BHS.show(WordResults2,start=39948,end=39948)

## 3.8 Center word - alternative 2<a class="anchor" id="bullet3x8"></a>
##### [Back to TOC](#TOC)

In [129]:
# Number of words in Torah (defined as items separeted by spaces) 
# following regexp (.+) selects for values of feature trailer that are 1 or more characters in length

WordQuery3 = '''
book book=Genesis|Exodus|Leviticus|Numeri|Deuteronomium

  word trailer~\ $
'''

WordResults3 = BHS.search(WordQuery3)

  0.60s 68434 results


In [130]:
# Just to check. query for maqafs

MaqafQuery4 = '''
book book=Genesis|Exodus|Leviticus|Numeri|Deuteronomium

  word trailer=&
'''

MaqafResults4 = BHS.search(MaqafQuery4)

  0.35s 11452 results


Check if the numbers do add up: 68434 (spaces) +  11452 (maqafs) =? 79886 (total)  YES!

Find the midpoint: 68434/2 = 34217

In [131]:
BHS.show(WordResults2,start=34217,end=34217)